substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -0,0 +1,290 @@
+import com.oracle.objectfile.debuginfo.DebugInfoProvider;
+import com.oracle.svm.hosted.image.sources.SourceManager;
+import com.oracle.svm.hosted.meta.HostedMethod;
+import com.oracle.svm.hosted.meta.HostedType;
+
+import jdk.vm.ci.code.site.Mark;
+import jdk.vm.ci.meta.LineNumberTable;
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+import jdk.vm.ci.meta.ResolvedJavaType;
+
+/**
+ * Implementation of the DebugInfoProvider API interface that allows type, code and heap data info
+ * to be passed to an ObjectFile when generation of debug info is enabled.
+ */
+class NativeImageDebugInfoProvider implements DebugInfoProvider {
+    private final DebugContext debugContext;
+    private final NativeImageCodeCache codeCache;
+    @SuppressWarnings("unused") private final NativeImageHeap heap;
+
+    NativeImageDebugInfoProvider(DebugContext debugContext, NativeImageCodeCache codeCache, NativeImageHeap heap) {
+        super();
+        this.debugContext = debugContext;
+        this.codeCache = codeCache;
+        this.heap = heap;
+    }
+
+    @Override
+    public Stream<DebugTypeInfo> typeInfoProvider() {
+        return Stream.empty();
+    }
+
+    @Override
+    public Stream<DebugCodeInfo> codeInfoProvider() {
+        return codeCache.compilations.entrySet().stream().map(entry -> new NativeImageDebugCodeInfo(entry.getKey(), entry.getValue()));
+    }
+
+    @Override
+    public Stream<DebugDataInfo> dataInfoProvider() {
+        return Stream.empty();
+    }
+
+    /**
+     * implementation of the DebugCodeInfo API interface that allows code info to be passed to an
+     * ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugCodeInfo implements DebugCodeInfo {
+        private final HostedMethod method;
+        private final ResolvedJavaType javaType;
+        private final CompilationResult compilation;
+        private Path fullFilePath;
+
+        NativeImageDebugCodeInfo(HostedMethod method, CompilationResult compilation) {
+            this.method = method;
+            HostedType declaringClass = method.getDeclaringClass();
+            Class<?> clazz = declaringClass.getJavaClass();
+            this.javaType = declaringClass.getWrapped();
+            this.compilation = compilation;
+            fullFilePath = ImageSingletons.lookup(SourceManager.class).findAndCacheSource(javaType, clazz);
+        }
+
+        @SuppressWarnings("try")
+        @Override
+        public void debugContext(Consumer<DebugContext> action) {
+            try (DebugContext.Scope s = debugContext.scope("DebugCodeInfo", method)) {
+                action.accept(debugContext);
+            } catch (Throwable e) {
+                throw debugContext.handle(e);
+            }
+        }
+
+        @Override
+        public String fileName() {
+            if (fullFilePath != null) {
+                return fullFilePath.getFileName().toString();
+            }
+            return "";
```

Comment:
This is actually a bugfix. (prevent NPE at `com.oracle.svm.hosted.image.NativeImageDebugInfoProvider.NativeImageDebugCodeInfo#lineInfoProvider`)
```java
if (fileName().length() == 0) { ... }
```